python/pyspark/sql/connect/client/retries.py

```diff
@@ -0,0 +1,228 @@
+        self._attempt = 0
+        self._next_wait: float = self._policy.initial_backoff
+
+    @property
+    def policy(self):
+        return self._policy
+
+    @property
+    def name(self):
+        return self.policy.name
+
+    def can_retry(self, exception: BaseException):
+        return self.policy.can_retry(exception)
+
+    def next_attempt(self) -> Optional[int]:
+        """
+        Returns
+        -------
+            Randomized time (in milliseconds) to wait until this attempt
+            or None if this policy doesn't allow more retries.
+        """
+
+        if self.policy.max_retries is not None and self._attempt >= self.policy.max_retries:
+            # No more retries under this policy
+            return None
+
+        self._attempt += 1
+        wait_time = self._next_wait
+
+        # Calculate future backoff
+        if self.policy.max_backoff is not None:
+            self._next_wait = min(
+                float(self.policy.max_backoff), wait_time * self.policy.backoff_multiplier
+            )
+
+        # Jitter current backoff, after the future backoff was computed
+        if wait_time >= self.policy.min_jitter_threshold:
+            wait_time += random.uniform(0, self.policy.jitter)
+
+        # Round to whole number of milliseconds
+        return int(wait_time)
+
+
+class AttemptManager:
+    """
+    Simple ContextManager that is used to capture the exception thrown inside the context.
+    """
+
+    def __init__(self, retrying: "Retrying") -> None:
+        self._retrying = retrying
+
+    def __enter__(self) -> None:
+        pass
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exception: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> Optional[bool]:
+        if isinstance(exception, BaseException):
+            # Swallow the exception.
+            if self._retrying.accept_exception(exception):
+                return True
+            # Bubble up the exception.
+            return False
+        else:
+            self._retrying.accept_succeeded()
+            return None
+
+
+class Retrying:
+    """
+    This class is a point of entry into the retry logic.
+    The class accepts a list of retry policies and applies them in given order.
+    The first policy accepting an exception will be used.
+    """
+
+    def __init__(
+        self,
+        policies: typing.Union[RetryPolicy, typing.Iterable[RetryPolicy]],
+        sleep: Callable[[float], None] = time.sleep,
+    ) -> None:
+        if isinstance(policies, RetryPolicy):
+            policies = [policies]
+        self._policies: List[RetryPolicyState] = [policy.to_state() for policy in policies]
+        self._sleep = sleep
+
+        self._exception: Optional[BaseException] = None
+        self._done = False
+
+    def can_retry(self, exception: BaseException) -> bool:
+        return any(policy.can_retry(exception) for policy in self._policies)
```

Comment:
different policies can have different attempt numbers.
by passing only `exception` how can a policy determine if it can retry, without e.g. knowing that this is the 10th attempt, while that policy allows only 4 attempts?

edit: I see, this will check if the exception can even possibly be retried, and the 
```
            wait_time = policy.next_attempt()
            if wait_time is not None:
```
will actually check if next attempt is allowed.

maybe add this to can_retry documentation, and a comment here, that it only checks if the exception can be retried here, and whether it has more retry attempts is checked later.

Reply:
This is a behavior I've implemented specifically to match the current design -- it also does classify errors into "retryable" and "not retryable", and the logic of spacing retries is separate.

I can add some more comments

Reply:
Yep, it matches the current design. Just took me a moment to understand so some additional comments might be useful.